sdk/src/signer/mod.rs

```diff
@@ -103,6 +109,25 @@ pub fn unique_signers(signers: Vec<&dyn Signer>) -> Vec<&dyn Signer> {
     signers.into_iter().unique_by(|s| s.pubkey()).collect()
 }
 
+/// The `EncodableKey` trait defines the interface by which cryptographic keys/keypairs are read,
+/// written, and derived from sources.
+pub trait EncodableKey: Sized {
+    fn pubkey_string(&self) -> Result<String, Box<dyn error::Error>>;
```

Comment:
could we replace this with an associated type like `type Pubkey: ToString;`

Reply:
That is a good idea! This way, the function will be a lot more general.

Reply:
Actually, for symmetric keys like `AeKey`, there is no `Pubkey` type. The function `pubkey_string(&self)` can return an error in this case like [here](https://github.com/solana-labs/solana/pull/30901/files#diff-8e0f5e2484afe07494baa6995029cfde17aaa6c3e7030c98f7a0d266790c4a66R222), but if we now have an associated `Pubkey` type in the trait, we have to define a dummy pubkey type for symmetric keys that will never be used, so I am not sure if this is the cleanest way to define things. Wdyt?

Reply:
sounds like maybe this attribute doesn't belong on the trait, then?

Reply:
Yeah, it seems like removing it is the way to go.

Reply:
were you still planning to remove `pubkey_string()`?

Reply:
I was planning to not have an associated `Pubkey` type, but still have the function `pubkey_string()` because it simplifies [confirming pubkeys](https://github.com/solana-labs/solana/pull/30947/files#diff-541bd0923c958e58e17e227c88be47279e4e25605979f0ad1fc536f63d4ae688R1121) in `encodable_key_from_seed_phrase` and I want `encodable_key_from_seed_phrase` to also support symmetric keys. So when symmetric keys implement `EncodableKey`, I was planning on returning a suitable error type like [this](https://github.com/solana-labs/solana/pull/30901/files#diff-8e0f5e2484afe07494baa6995029cfde17aaa6c3e7030c98f7a0d266790c4a66R222).

The canonical way to make `encodable_key_from_seed_phrase` to work is to probably have two traits, `EncodableKey` (with `pubkey_string` removed) and some general `CryptographicKey` trait that has an associated function `pubkey_string() -> Option<String>` or something like that. Then have `encodable_key_from_seed_phrase<K: EncodableKey + CryptographicKey>(...) -> Result<K, Error>`. But I feel like this is more trouble than it is worth.

Reply:
if the confirm thing is all that `pubkey_string` is helping, how about we take it out of `encodable_key_from_path` altogether?  if we want `encodable_key_from_path` to be generic on symmetric keys, even the mention of a pubkey is confusing. would it be sufficient to move the `if confirm_pubkey` block out to a helper and call that on the result of `encodable_key_from_path` optionally from `keypair_from_path` and `keypair_from_seed_phrase`?

Reply:
Oh yes, I don't know how I didn't think of just removing the pubkey part out of the generic helper function. I guess I was too focused on minimizing changes to the existing code. I will remove `pubkey_string`.